packages/sanity/src/core/store/_legacy/history/useTimelineController.ts

```diff
@@ -0,0 +1,225 @@
+  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
+  const [timelineError, setTimelineError] = useState<Error | null>(null)
+
+  /**
+   * The mutable TimelineController, used internally
+   */
+  const controller = useMemo(
+    () =>
+      historyStore.getTimelineController({
+        client,
+        documentId,
+        documentType,
+      }),
+    [client, documentId, documentType, historyStore]
+  )
+
+  /**
+   * A BehaviorSubject which can be subscribed by multiple observers, which broadcasts
+   * the the latest state of the (mutable) TimelineController.
+   */
+  const timelineController$ = useMemo(
+    () => new BehaviorSubject<TimelineController>(controller),
+    [controller]
+  )
+
+  /**
+   * Broadcasts the updated state on TimelineController's `handler` updates (as well as on mount / unmount).
+   *
+   * Note that TimelineController triggers updates very frequently when fetching transactions, which it does so
+   * in fairly small batches.
+   *
+   * This will also suspend TimelineController transaction fetching on cleanup, ensuring that fetches don't
+   * continue once the document pane has been closed.
+   */
+  useEffect(() => {
+    controller.setRange(since || null, rev || null)
+    timelineController$.next(controller)
+
+    controller.handler = (err, innerController) => {
+      if (err) {
+        timelineController$.error(err)
+      } else {
+        /**
+         * NOTE: TimelineController requires that you call `setRange` manually whenever its internal
+         * timeline has changed (e.g. has a result of fetched transactions).
+         * Ideally, TimelineController would be updated to handle this automatically for us.
+         * Until then, this workaround waits until the next call stack before calling `setRange`.
+         */
+        setTimeout(() => {
+          innerController.setRange(since || null, rev || null)
+          timelineController$.next(innerController)
+        }, 0)
+      }
+    }
+    controller.resume()
+    return () => controller.suspend()
+  }, [rev, since, controller, timelineController$])
+
+  /**
+   * Fetch document snapshots and update the mutable controller.
+   * Unsubscribes on clean up, preventing double fetches in strict mode.
+   */
+  useEffect(() => {
+    if (!snapshotsSubscriptionRef.current) {
+      snapshotsSubscriptionRef.current = remoteSnapshots(
+        client,
+        {draftId: `drafts.${documentId}`, publishedId: documentId},
+        documentType
+      ).subscribe((ev: RemoteSnapshotVersionEvent) => {
+        controller.handleRemoteMutation(ev)
+      })
+    }
+    return () => {
+      if (snapshotsSubscriptionRef.current) {
+        snapshotsSubscriptionRef.current.unsubscribe()
+        snapshotsSubscriptionRef.current = null
+      }
+    }
+  }, [client, controller, documentId, documentType])
+
+  /**
+   * Custom hook which wraps around `useSyncExternalStore`. Accepts a selector function which can be used
```

Comment:
I have a bad feeling about hooks that create hooks that are then called. It's risky magic and can easily break with React concurrent features.

I'd encourage you to consider a 2 hooks pattern where you instead pass the store:

```tsx
import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/with-selector'

function Component() {}
  const {timelineStore} = useDocumentPane()
  const onOlderRevision = useTimelineSelector(timelineStore, (state) => state.onOlderRevision)
}

function useTimelineSelector(timelineStore, selector, isEqual?) {
  return useSyncExternalStoreWithSelector(timelineStore.subscribe, timelineStore.getSnapshot, timelineStore.getServerSnapshot, selector, isEqual)
}
```

We also ought to stop putting more stuff on the already giant `DocumentPaneContext` and start splitting it into multiple, separate, context providers if we are to properly solve todays performance bottlenecks in the Studio.
The pattern above can be ugpraded to the multi-context pattern:
```tsx
import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/with-selector'

function Component() {}
  const onOlderRevision = useTimelineSelector((state) => state.onOlderRevision)
}

function useTimelineSelector(selector, isEqual?) {
  const timelineStore = useTimelineStore()
  return useSyncExternalStoreWithSelector(timelineStore.subscribe, timelineStore.getSnapshot, timelineStore.getServerSnapshot, selector, isEqual)
}

function useTimelineStore() {
  return useContext(TimelineStoreContext)
}
```

Reply:
Agree with cody about the pattern of hooks creating hooks.

We are well aware that we need to set aside time to split up the DocumentPaneProvider/context, but don't necessarily think it should get in the way of merging this PR unless there's evidence it has any additional negative impact on perf?

Reply:
@stipsan Really appreciate the feedback here

> I'd encourage you to consider a 2 hooks pattern where you instead pass the store:

Much better – that's been updated. We no longer have a hook returning a hook, nor do we store one in DocumentPaneProvider.

> We also ought to stop putting more stuff on the already giant DocumentPaneContext and start splitting it into multiple, separate, context providers

Completely agree and totally on board with this. I've opted to keep it as-is _for now_, but I think there's definitely an argument for breaking this up and we could look to do this holistically, provided that DX give the green light.

Some other notes:
- We create a `timelineStore` once in `DocumentPaneProvider` and store that in context. Components down the chain can retrieve that store to either subscribe to changes via `useTimelineSelector`, or alternatively call select methods which are called on the underlying `TimelineController` (mainly, obtaining ranges and loading more transactions)
- I think this is still an improvement over what was happening prior (storing the entire `TimelineController` class instance in context). But as you've suggested, a dedicated Timeline provider / context could go a long way in cleaning this up.
- Do we need `use-sync-external-store` if we're only supporting React 18?